ouroboros-consensus/src/Ouroboros/Consensus/Protocol/TPraos.hs

```diff
@@ -0,0 +1,359 @@
+          -- First, check whether we're in the overlay schedule
+          case (Map.lookup (convertSlot slot) $ tpraosLedgerViewOverlaySchedule lv) of
+            Nothing -> return $
+              -- Slot isn't in the overlay schedule, so we're in Praos
+              if fromIntegral (certifiedNatural y) < t
+                  then Just TPraosProof {
+                          tpraosEta       = coerce rho
+                        , tpraosLeader    = coerce y
+                        , tpraosProofSlot = slot
+                        , tpraosIsCoreNode = icn
+                        }
+                  else Nothing
+            Just Nothing ->
+              -- This is a non-active slot; nobody may produce a block
+              return Nothing
+            Just (Just gkhash) ->
+              -- The given genesis key has authority to produce a block in this
+              -- slot. Check whether we're its delegate.
+              let GenDelegs dlgMap = tpraosLedgerViewDelegationMap lv
+              in do
+                let verKey = ocertVkCold tpraosIsCoreNodeOpCert
+                return $ case Map.lookup gkhash dlgMap of
+                  Just dlgHash | dlgHash == hashKey verKey ->
+                    Just TPraosProof
+                        { tpraosEta = coerce rho
+                          -- Note that this leader value is not checked for
+                          -- slots in the overlay schedule, so we could set it
+                          -- to whatever we want. We evaluate it as normal for
+                          -- simplicity's sake.
+                        , tpraosLeader = coerce y
+                        , tpraosProofSlot = slot
+                        , tpraosIsCoreNode = icn
+                        }
+                  _ -> Nothing
+
+  applyChainState TPraosNodeConfig{..} lv b cs = do
+    let slot = blockSlot b
+        SecurityParam (fromIntegral -> k) = tpraosSecurityParam tpraosParams
+
+    newCS <- ExceptT . return $ applySTS @(PRTCL c)
+      $ TRC ( STS.PrtclEnv
+                ( STS.OverlayEnv
+                  (tpraosLedgerViewProtParams lv)
+                  (tpraosLedgerViewOverlaySchedule lv)
+                  (tpraosLedgerViewEpochNonce lv)
+                  (tpraosLedgerViewPoolDistr lv)
+                  (tpraosLedgerViewDelegationMap lv)
+                )
+                (convertSlot slot)
+            , ChainState.toPRTCLState cs
+            , headerToBHeader (Proxy :: Proxy (TPraos c)) b
+            )
+
+    return . ChainState.prune k $ ChainState.appendState newCS cs
+
+  -- Rewind the chain state
+  --
+  -- We don't roll back to the exact slot since that slot might not have been
+  -- filled; instead we roll back the the block just before it.
+  rewindChainState TPraosNodeConfig{..} cs rewindTo = ChainState.rewind rewindTo cs
+
+  -- NOTE: We redefine `preferCandidate` but NOT `compareCandidates`
+  -- NOTE: See note regarding clock skew.
```

Comment:
TODO: Chain selection in TPraos should take certificate issue numbers into account. (But this will require a generalization of `OuroborosTag`. Coming up.)